### PR TITLE
[release/7.0] Fix regression in `JsonIgnoreAttribute` handling of unsupported property types

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Backport of #76828 to release/7.0

## Customer Impact

We received a [couple of](https://github.com/dotnet/runtime/issues/76807) [customer reports](https://github.com/dotnet/runtime/issues/76794) that `JsonIgnoreAttribute` handling when applied to unsupported property types has regressed in .NET 7. Even though there are workarounds for the issue, these are generally unintuitive to users and require code modifications, which might inhibit migration of existing applications to .NET 7.

## Testing

Added regression testing for the identified scenarios.

## Risk

Low. Makes a minor change in product code so that a lazy loading mechanism already employed in source generated metadata is also applied to reflection derived metadata.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs [hunk 1]
```diff
@@ -37,11 +32,17 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
...
-        private void AddPropertiesAndParametersUsingReflection()
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:RequiresUnreferencedCode",
+            Justification = "The ctor is marked RequiresUnreferencedCode.")]
+        internal override void LateAddProperties()
         {
-            Debug.Assert(PropertyInfoForTypeInfo.ConverterStrategy == ConverterStrategy.Object);
+            Debug.Assert(!IsConfigured);
+            Debug.Assert(PropertyCache is null);
+
+            if (Kind != JsonTypeInfoKind.Object)
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:42]: Is it even possible we got here when it's not Object? IIRC this method is never called for non Objects

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs [hunk 2]
```diff
@@ -37,11 +32,17 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
             converter.ConfigureJsonTypeInfoUsingReflection(this, options);
         }
 
-        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
-        private void AddPropertiesAndParametersUsingReflection()
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:RequiresUnreferencedCode",
+            Justification = "The ctor is marked RequiresUnreferencedCode.")]
+        internal override void LateAddProperties()
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:37]: why do we need to change visibility? We don't seem to change usage of this method. Can you explain how this fix is fixing the customer scenario?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs [hunk 3]
```diff
@@ -37,11 +32,17 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
             converter.ConfigureJsonTypeInfoUsingReflection(this, options);
         }
 
-        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:40]: is this change relevant to actual fix?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs [hunk 4]
```diff
@@ -37,11 +32,17 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
...
+            Justification = "The ctor is marked RequiresUnreferencedCode.")]
+        internal override void LateAddProperties()
         {
-            Debug.Assert(PropertyInfoForTypeInfo.ConverterStrategy == ConverterStrategy.Object);
+            Debug.Assert(!IsConfigured);
+            Debug.Assert(PropertyCache is null);
+
+            if (Kind != JsonTypeInfoKind.Object)
+            {
+                return;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:44]: Honestly we should not be calling this method for non-objects. Fine for servicing fix but if we need to call it for non-objects we might have a bug elsewhere (I'm guessing only JsonTypeInfo.Properties might be calling this incorrectly and will be more efficient to make it return empty list for that case than go here)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:42]: Looks like we are: https://github.com/dotnet/runtime/pull/76828#discussion_r991550038

Ideally this shouldn't happen, but rather than introducing more churn I'm replicating what the source gen override of the same method is doing.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:37]: Because that's the accessibility of `LateAddProperties`, which predates this change.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:40]: Because the logic has been moved to `LateAddProperties` which is a virtual method, linker/AOT annotations must match that of the base method (which doesn't have any). In this case it safe to suppress warnings since it's an instance method on a type with annotated constructors.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:44]: See https://github.com/dotnet/runtime/pull/76869#discussion_r992411365

**Review by jeffhandley — APPROVED** [review]
Thanks for getting this resolved. It was approved by tactics for inclusion in GA.
